lib/Sema/CSSimplify.cpp

```diff
@@ -4239,6 +4238,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+    // If this is an archetype type, check if any types it conforms to
+    // (superclass or protocols) have the attribute.
+    if (auto archetype = dyn_cast<ArchetypeType>(canType)) {
+      SmallVector<Type, 2> componentTypes;
+      for (auto protocolDecl : archetype->getConformsTo())
+        componentTypes.push_back(protocolDecl->getDeclaredType());
+      if (auto superclass = archetype->getSuperclass())
+        componentTypes.push_back(superclass);
+      return calculateForComponentTypes(componentTypes);
+    }
+
+    // If this is a protocol composition, check if any of its members have the
+    // attribute.
+    if (auto protocolComp = dyn_cast<ProtocolCompositionType>(canType))
+      return calculateForComponentTypes(protocolComp->getMembers());
+
+    // Otherwise, this must be a nominal type.
+    // Dynamic calling doesn't work for tuples, etc.
+    auto nominal = canType->getAnyNominal();
+    if (!nominal) return DynamicCallableMethods();
+
+    // If this type conforms to a protocol which has the attribute, then
+    // look up the methods.
+    for (auto p : nominal->getAllProtocols())
+      if (p->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
+
+    // Walk superclasses, if present.
+    llvm::SmallPtrSet<const NominalTypeDecl*, 8> visitedDecls;
+    while (1) {
+      // If we found a circular parent class chain, reject this.
+      if (!visitedDecls.insert(nominal).second)
+        return DynamicCallableMethods();
+
+      // If this type has the attribute on it, then look up the methods.
+      if (nominal->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
+
+      // If this type is a class with a superclass, check superclasses.
+      if (auto *cd = dyn_cast<ClassDecl>(nominal)) {
+        if (auto superClass = cd->getSuperclassDecl()) {
+          nominal = superClass;
+          continue;
+        }
+      }
+
+      return DynamicCallableMethods();
+    }
+  };
+
+  return CS.DynamicCallableCache[canType] = calculate();
+}
+
+// Simplify applicable function constraint for an application of a
+// @dynamicCallable type.
+static ConstraintSystem::SolutionKind
+simplifyDynamicCallableApplicableFnConstraint(
+    ConstraintSystem &CS, Type type1, Type type2,
+    ConstraintLocatorBuilder locator,
+    ConstraintSystem::TypeMatchOptions flags) {
+  using SolutionKind = ConstraintSystem::SolutionKind;
+
+  bool error = false;
+  auto methods = getDynamicCallableMethods(type2, CS, locator, error);
+  if (error || !methods.isValid()) return SolutionKind::Error;
+
+  auto &ctx = CS.getASTContext();
+  auto func1 = type1->castTo<FunctionType>();
+
+  // Determine whether to call the positional arguments method or the
+  // keyword arguments method.
+  bool useKwargsMethod = methods.argumentsMethod == nullptr;
+  if (!useKwargsMethod) {
+    for (auto param : func1->getParams()) {
```

Comment:
Nit: you can simplify this loop into: 

```c++
llvm::find_if(func1->getParams(), [](AnyFunctionType::Param param) { return param.hasLabel(); });
```

Reply:
Thanks! Done in https://github.com/apple/swift/pull/16980/commits/a982340e8d3494dc6c7924b63c2f6f2da1353142.

Reply:
Actually, this change caused `attr/attr_dynamic_callable.swift` to fail for some reason.
`llvm::find_if` produces different behavior than the original for-loop code.

@xedin, do you have any idea what's going on?